Utils/FaviconFetcher.swift

```diff
@@ -0,0 +1,130 @@
+    private var _favicons = [Favicon]() // An internal cache of favicons found for this url
+
+    class func getForUrl(url: NSURL, profile: Profile, callback: ([Favicon]) -> Void) {
+        dispatch_async(queue) {
+            let f = FaviconFetcher(url: url)
+            f.loadFavicons(profile, callback: callback)
+        }
+    }
+
+    private init(url: NSURL) {
+        siteUrl = url
+    }
+
+    private func loadFavicons(profile: Profile, callback: ([Favicon]) -> Void) {
+        if _favicons.count == 0 {
+            // Initially look for tags in the page
+            loadFromDoc()
+        }
+
+        // If that didn't find anything, look for a favicon.ico for this host
+        if _favicons.count == 0 {
+            loadFromHost()
+        }
+
+        var filledCount = 0
+        for (i, icon) in enumerate(_favicons) {
+            getFavicon(icon, profile: profile) { icon in
+                if let icon = icon {
+                    self._favicons[i] = icon
+                }
+
+                filledCount++
+                if filledCount == self._favicons.count {
+                    self._favicons.sort({ (a, b) -> Bool in
+                        return a.width > b.width
+                    })
+
+                    dispatch_async(dispatch_get_main_queue()) {
+                        callback(self._favicons)
+                    }
+                }
+            }
+        }
+    }
+
+    // Loads favicon.ico on the host domain for this url
+    private func loadFromHost() {
+        var url = NSURL(scheme: siteUrl.scheme!, host: siteUrl.host, path: "/favicon.ico")
+        var request = NSMutableURLRequest(URL: url!)
+        request.HTTPMethod = "HEAD"
+        request.timeoutInterval = 5
+
+        var response: NSURLResponse? = nil;
+        var err = NSErrorPointer()
+        NSURLConnection.sendSynchronousRequest(request, returningResponse: &response, error: err)
+        if let httpResponse = response as? NSHTTPURLResponse {
+            if httpResponse.statusCode == 200 {
+                let icon = Favicon(url: url!.absoluteString!, type: IconType.Guess)
+                _favicons.append(icon)
+            }
+        }
+    }
+
+    // Loads and parses an html document and tries to find any known favicon-type tags for the page
+    private func loadFromDoc() {
+        var err: NSError?
+
+        if let data = NSData(contentsOfURL: siteUrl),
+           let element = RXMLElement(fromHTMLData: data) {
+            element.iterate("head.meta") { meta in
+                if let refresh = meta.attribute("http-equiv"),
+                   let content = meta.attribute("content"),
+                   let index = content.rangeOfString("URL="),
+                   let url = NSURL(string: content.substringFromIndex(advance(index.startIndex,4))) {
+                    if refresh == "Refresh" {
+                        self.siteUrl = url
+                        return self.loadFromDoc()
```

Comment:
Nit: This method returns Void but we're calling the method from the return statement. Can we move this up just to make it more clear?